modules/imgproc/src/corner.avx.cpp

```diff
@@ -49,132 +49,112 @@
-
-    __m256 v20 = _mm256_unpacklo_ps(v14, v15);           // b0, b1, b2, b3, c0, c1, c2, c3,
-    __m256 v21 = _mm256_unpackhi_ps(v14, v15);           // b4, b5, b6, b7, c4, c5, c6, c7,
-    __m256 v22 = _mm256_unpacklo_ps(v10, v13);           // a0, a1, a2, a3, x,  x,  x,  x,
-    __m256 v23 = _mm256_unpackhi_ps(v10, v13);           // a4, a5, a6, a7, x,  x,  x,  x,
-
-    a = _mm256_permute2f128_ps(v22, v23, 0x20);          // a0, a1, a2, a3, a4, a5, a6, a7,
-    b = _mm256_permute2f128_ps(v20, v21, 0x20);          // b0, b1, b2, b3, b4, b5, b6, b7,
-    c = _mm256_permute2f128_ps(v20, v21, 0x31);          // c0, c1, c2, c3, c4, c5, c6, c7,
-}
-
-// realign four 3-packed vector to three 4-packed vector
-static void v_pack4x3to3x4(const __m128i& s0, const __m128i& s1, const __m128i& s2, const __m128i& s3, __m128i& d0, __m128i& d1, __m128i& d2)
-{
-    d0 = _mm_or_si128(s0, _mm_slli_si128(s1, 12));
-    d1 = _mm_or_si128(_mm_srli_si128(s1, 4), _mm_slli_si128(s2, 8));
-    d2 = _mm_or_si128(_mm_srli_si128(s2, 8), _mm_slli_si128(s3, 4));
-}
-
-// separate high and low 128 bit and cast to __m128i
-static void v_separate_lo_hi(const __m256& src, __m128i& lo, __m128i& hi)
-{
-    lo = _mm_castps_si128(_mm256_castps256_ps128(src));
-    hi = _mm_castps_si128(_mm256_extractf128_ps(src, 1));
-}
-
-// interleave three 8-float vector and store
-static void store_interleave(float* ptr, const __m256& a, const __m256& b, const __m256& c)
-{
-    __m128i a0, a1, b0, b1, c0, c1;
-    v_separate_lo_hi(a, a0, a1);
-    v_separate_lo_hi(b, b0, b1);
-    v_separate_lo_hi(c, c0, c1);
-
-    v_uint32x4 z = v_setzero_u32();
-    v_uint32x4 u0, u1, u2, u3;
-    v_transpose4x4(v_uint32x4(a0), v_uint32x4(b0), v_uint32x4(c0), z, u0, u1, u2, u3);
-    v_pack4x3to3x4(u0.val, u1.val, u2.val, u3.val, a0, b0, c0);
-    v_transpose4x4(v_uint32x4(a1), v_uint32x4(b1), v_uint32x4(c1), z, u0, u1, u2, u3);
-    v_pack4x3to3x4(u0.val, u1.val, u2.val, u3.val, a1, b1, c1);
-
-#if !defined(__GNUC__) || defined(__INTEL_COMPILER)
-    _mm256_storeu_ps(ptr,      _mm256_castsi256_ps(_mm256_setr_m128i(a0, b0)));
-    _mm256_storeu_ps(ptr + 8,  _mm256_castsi256_ps(_mm256_setr_m128i(c0, a1)));
-    _mm256_storeu_ps(ptr + 16, _mm256_castsi256_ps(_mm256_setr_m128i(b1, c1)));
-#else
-    // GCC: workaround for missing AVX intrinsic: "_mm256_setr_m128()"
-    _mm256_storeu_ps(ptr, _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(a0)), _mm_castsi128_ps(b0), 1));
-    _mm256_storeu_ps(ptr + 8, _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(c0)), _mm_castsi128_ps(a1), 1));
-    _mm256_storeu_ps(ptr + 16,  _mm256_insertf128_ps(_mm256_castps128_ps256(_mm_castsi128_ps(b1)), _mm_castsi128_ps(c1), 1));
-#endif
-}
-
-int calcMinEigenValLine_AVX(const float* cov, float* dst, int width)
+int calcMinEigenValLine_AVX(const float* cov_x2, const float* cov_xy, const float* cov_y2, float* dst, int width)
```

Comment:
Please note, than this file is named with `.avx.cpp` suffix. It should contain AVX optimization only. There is no place to support AVX512.

Generic support for different instruction sets are implemented through:
- `.simd.hpp`
- and `.dispatch.cpp`

files.

See also: https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

Reply:
Thanks for the info @alalek 

> Please note, than this file is named with .avx.cpp suffix. It should contain AVX optimization only. There is no place to support AVX512.

I had actually worked this out and also found that by moving the AVX512 code to a file with extension `avx_512f.cpp` it could compile correctly with `-mavx512f` build flags etc. (Not sure if this intended because currently there are no source files in OpenCV with that file extension but it is possible - https://github.com/komakai/opencv/commit/f71fa77a4c22ac46dea836d240d3685460d747aa)

As per my previous comment - basically having 3 separate copies of the same algorithm doesn’t make much sense so I am continuing to try and find a way to make a single generic version like https://github.com/komakai/opencv/commit/cfbfea0966b0ff241d3772190edfbe9153908eb1

Unfortunately because only AVX2 is supported through OpenCV SIMD intrinsics currently, this actually results in worse performance on a machine with AVX but not AVX2. I guess the thing to do would be to add AVX support to OpenCV SIMD intrinsics - shouldn’t be too difficult - correct?